core/src/poh_recorder.rs

```diff
@@ -135,39 +127,41 @@ impl PohRecorder {
 
     /// returns if leader tick has reached, how many grace ticks were afforded,
     ///   imputed leader_slot and self.start_slot
+    /// reached_leader_tick() == true means "ready for a bank"
     pub fn reached_leader_tick(&self) -> (bool, u64, Slot, Slot) {
-        let slot =
-            leader_schedule_utils::tick_height_to_slot(self.ticks_per_slot, self.tick_height);
-
         trace!(
-            "tick_height {}, start_tick {} start_leader_at_tick {:?}, grace {}",
+            "tick_height {}, start_tick {}, start_leader_at_tick {:?}, grace {}, has_bank {}",
             self.tick_height,
             self.start_tick,
             self.start_leader_at_tick,
-            self.max_last_leader_grace_ticks
+            self.grace_ticks,
+            self.has_bank()
         );
 
+        let next_tick = self.tick_height + 1;
+
         if let Some(target_tick) = self.start_leader_at_tick {
-            let leader_ideal_start_tick =
-                target_tick.saturating_sub(self.max_last_leader_grace_ticks);
-            // Check if either grace period has expired,
-            // or target tick is = grace period (i.e. poh recorder was just reset)
-            if self.tick_height >= target_tick
-                || self.max_last_leader_grace_ticks >= target_tick.saturating_sub(self.start_tick)
+            // if we've reached the target tick OR poh was *just* reset
+            if next_tick >= target_tick
```

Comment:
what happens if we overrun the slot, i.e. next_tick jumped past the scheduled `target_tick` slot?

Reply:
we're still ready for a bank, assuming the imputed slot is ours as a leader

this was the big breakthrough from my last "groom"

Reply:
replay_stage checks to see if we actually should be leading